File: trip-planner-web/src/api/trip.ts
import api from './index'

export interface TripCreateReq {
  tripName: string
  destination: string
  startDate: string
  endDate: string
  numPeople?: number
  notes?: string
  totalBudget?: number
}

export interface TripUpdateReq {
  tripName?: string
  destination?: string
  startDate?: string
  endDate?: string
  numPeople?: number
  notes?: string
  totalBudget?: number
}

export interface TripData {
  id: number
  tripName: string
  destination: string
  startDate: string
  endDate: string
  numPeople: number
  notes: string
  status: string
  totalBudget: number
  coverImageUrl: string
  daysCount: number
  createTime: string
  updateTime: string
}

export interface PageResult<T> {
  total: number
  page: number
  size: number
  records: T[]
}

export interface TripQuery {
  page?: number
  size?: number
  status?: string
  keyword?: string
  sortBy?: string
  sortDir?: string
}

export const tripApi = {
  create(data: TripCreateReq): Promise<TripData> {
    return api.post('/trips', data)
  },
  list(query: TripQuery): Promise<PageResult<TripData>> {
    return api.get('/trips', { params: query })
  },
  getById(id: number): Promise<TripData> {
    return api.get(`/trips/${id}`)
  },
  update(id: number, data: TripUpdateReq): Promise<TripData> {
    return api.put(`/trips/${id}`, data)
  },
  delete(id: number): Promise<void> {
    return api.delete(`/trips/${id}`)
  },
  copy(id: number): Promise<TripData> {
    return api.post(`/trips/${id}/copy`)
  },
  updateStatus(id: number, status: string): Promise<TripData> {
    return api.patch(`/trips/${id}/status`, { status })
  }
}

File: trip-planner-flutter/lib/models/trip.dart
class Trip {
  final int id;
  final String tripName;
  final String destination;
  final String? startDate;
  final String? endDate;
  final int? numPeople;
  final String? notes;
  final String status;
  final double? totalBudget;
  final String? coverImageUrl;
  final int? daysCount;
  final String? createTime;
  final String? updateTime;

  Trip({
    required this.id, required this.tripName, required this.destination,
    this.startDate, this.endDate, this.numPeople, this.notes,
    required this.status, this.totalBudget, this.coverImageUrl,
    this.daysCount, this.createTime, this.updateTime,
  });

  factory Trip.fromJson(Map<String, dynamic> json) => Trip(
    id: json['id'] as int,
    tripName: json['tripName'] as String,
    destination: json['destination'] as String,
    startDate: json['startDate'] as String?,
    endDate: json['endDate'] as String?,
    numPeople: json['numPeople'] as int?,
    notes: json['notes'] as String?,
    status: (json['status'] as String?) ?? 'PLANNING',
    totalBudget: (json['totalBudget'] as num?)?.toDouble(),
    coverImageUrl: json['coverImageUrl'] as String?,
    daysCount: json['daysCount'] as int?,
    createTime: json['createTime'] as String?,
    updateTime: json['updateTime'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id, 'tripName': tripName, 'destination': destination,
    'startDate': startDate, 'endDate': endDate, 'numPeople': numPeople,
    'notes': notes, 'status': status, 'totalBudget': totalBudget,
    'coverImageUrl': coverImageUrl, 'daysCount': daysCount,
  };
}

class TripCreateReq {
  final String tripName;
  final String destination;
  final String? startDate;
  final String? endDate;
  final int? numPeople;
  final String? notes;
  final double? totalBudget;

  TripCreateReq({
    required this.tripName, required this.destination,
    this.startDate, this.endDate, this.numPeople, this.notes, this.totalBudget,
  });

  Map<String, dynamic> toJson() => {
    'tripName': tripName, 'destination': destination,
    if (startDate != null) 'startDate': startDate,
    if (endDate != null) 'endDate': endDate,
    if (numPeople != null) 'numPeople': numPeople,
    if (notes != null) 'notes': notes,
    if (totalBudget != null) 'totalBudget': totalBudget,
  };
}

class TripUpdateReq {
  final String? tripName;
  final String? destination;
  final String? startDate;
  final String? endDate;
  final int? numPeople;
  final String? notes;
  final double? totalBudget;

  TripUpdateReq({this.tripName, this.destination, this.startDate, this.endDate, this.numPeople, this.notes, this.totalBudget});

  Map<String, dynamic> toJson() => {
    if (tripName != null) 'tripName': tripName,
    if (destination != null) 'destination': destination,
    if (startDate != null) 'startDate': startDate,
    if (endDate != null) 'endDate': endDate,
    if (numPeople != null) 'numPeople': numPeople,
    if (notes != null) 'notes': notes,
    if (totalBudget != null) 'totalBudget': totalBudget,
  };
}

class PageResult<T> {
  final int total;
  final int page;
  final int size;
  final List<T> records;

  PageResult({required this.total, required this.page, required this.size, required this.records});

  factory PageResult.fromJson(Map<String, dynamic> json, T Function(Map<String, dynamic>) fromItem) {
    return PageResult(
      total: (json['total'] as num?)?.toInt() ?? 0,
      page: (json['page'] as num?)?.toInt() ?? 1,
      size: (json['size'] as num?)?.toInt() ?? 10,
      records: (json['records'] as List<dynamic>?)
          ?.map((e) => fromItem(e as Map<String, dynamic>))
          .toList() ?? [],
    );
  }
}
